### PR TITLE
fix(attention): materialize block_valid_mask for padded CUDA-graph prefill launches

### LINKED ISSUE #4002: [Bug] FA2 CUDA-graph no-split plans launch padded CTAs over uninitialized scheduler entries
## Description

FA2 tensor-core decode can launch CUDA-graph blocks beyond the scheduler entries initialized by the planner when all of these hold:

- `enable_cuda_graph=True`
- `disable_split_kv=True`
- tensor-core FA2 path
- a uniform decode query length

The scheduler constructs entries only for real request, query-tile, and KV-tile work, then pads `padded_batch_size` with the split-KV occupancy target. Since split-KV is disabled, the planner neither initializes the padded scheduler tail nor creates `block_valid_mask`. The FA2 kernel uses `padded_batch_size` directly as grid-x.

The same mechanics remain on current main at `71d31b5a23a3c0394edb36330dec1ce2a0def365`:

- [One-token decode passes `uniform_q_len=0`](https://github.com/flashinfer-ai/flashinfer/blob/71d31b5a23a3c0394edb36330dec1ce2a0def365/flashinfer/decode.py#L1215-L1241)
- [The scheduler creates real entries and then pads grid-x](https://github.com/flashinfer-ai/flashinfer/blob/71d31b5a23a3c0394edb36330dec1ce2a0def365/include/flashinfer/attention/scheduler.cuh#L621-L680)
- [Validity storage is allocated only for split-KV](https://github.com/flashinfer-ai/flashinfer/blob/71d31b5a23a3c0394edb36330dec1ce2a0def365/include/flashinfer/attention/scheduler.cuh#L855-L878)
- [The kernel launches `padded_batch_size` blocks in grid-x](https://github.com/flashinfer-ai/flashinfer/blob/71d31b5a23a3c0394edb36330dec1ce2a0def365/include/flashinfer/attention/prefill.cuh#L4150-L4172)

## Reproduction and evidence

Tested on an H100 80GB SXM5 with 132 SMs, batch size 2, 16 query heads, 2 KV heads, head dimension 128, page size 1, one query row per request, and split-KV disabled.

The planner returned grid-x 132 although it initialized only two scheduler entries. Filling the page-locked integer workspace with `0xA5` before planning left the first tail entries at `0xA5A5A5A5`. The poisoned plan was inspected and intentionally not launched.

With a controlled zeroed tail, Nsight captured grid `(132, 1, 2)`, or 264 CTAs, while the exact initialized grid is `(2, 1, 2)`, or 4 CTAs. This demonstrates redundant execution. Reused nonzero workspace contents create a conditional stale-index correctness risk; this report does not claim unconditional output corruption.

For a uniform decode query length, the exact graph-stable grid is:

`batch_size * ceil((q_len_per_request * num_qo_heads / num_kv_heads) / cta_tile_q)`

## Expected behavior

A no-split CUDA-graph plan must not launch blocks whose scheduler entries were never initialized. When query length is a checked graph invariant, the planner can use the exact initialized grid. Ragged and split-KV paths should retain their current behavior.

## Proposed fixes

- Root planner fix: [flashinfer-ai/flashinfer#4794](https://github.com/flashinfer-ai/flashinfer/pull/4794)
- Guarded integration fix for SGLang's pinned FlashInfer release: [sgl-project/sglang#36785](https://github.com/sgl-project/sglang/pull/36785)

### PR DESCRIPTION
**Pull request by gf239**
## 📌 Description

With CUDA graphs on and split-KV off, FA2 prefill can crash with `illegal memory access`.

Under CUDA graphs the kernel launches more blocks than the batch needs.
The extra blocks must be masked off.
The mask was only created when split-KV was on.
Without it the extra blocks read plan entries that were never written.
If the workspace still holds an earlier plan, they read out of bounds.

Who hits it:
- SGLang with `--enable-deterministic-inference` (tensor-core decode uses this planner).
- NVFP4 KV cache on Blackwell, where FlashInfer turns split-KV off itself.

Reproduced on an RTX 6000 Ada with the calls SGLang makes (16 requests, SGLang itself not run).
Fresh workspace: output happens to be correct.
Workspace reused from an earlier plan: crash.
With this PR both are correct.

Fix: create the mask whenever CUDA graphs are on.
Pass it to the kernel in every run function that uses the plan: ragged, paged, POD, batch POD.
Cost: `padded_batch_size` bytes of int workspace.
Plans without CUDA graphs no longer allocate the unused mask.

## 🔍 Related Issues

- Fixes #4002.
- #4794 uses an exact grid when all requests have the same query length. With it applied, the two new prefill tests still fail. The new decode test no longer pads, so that case #4794 covers.
- sgl-project/sglang#36785 works around the same crash in SGLang.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

| new test | main | with this PR |
| --- | --- | --- |
| `test_batch_prefill_cuda_graph_padding_without_split_kv[paged]` | fails | passes |
| `test_batch_prefill_cuda_graph_padding_without_split_kv[ragged]` | fails | passes |
| `test_tensor_core_decode_cuda_graph_padding_without_split_kv` | fails | passes |

Same on RTX 4090 and RTX 3080.
A fourth new test checks that a padded no-split CUDA-graph plan fits the buffers `workspace_size` reports.

On RTX 4090, the prefill, decode, POD and workspace-size test files give the same per-test results with and without this PR.

## 🔬 Experimental Track

- [ ] This PR is **experimental**: it adds or changes code under `flashinfer/experimental/` and/or an `@flashinfer_experimental_api`. Tracking issue: #
  - [ ] The tracking issue names an owner, the reason for the experimental path, and a graduation plan with a target release.
  - [ ] Core changes are limited to a thin entry point (signature, shared validation, feature-gate check, backend selection, handoff).
  - [ ] Tests live in `tests/experimental/` and were validated on the intended hardware; a runnable example is included.
  - [ ] Nothing is registered in `flashinfer/aot.py`, and no experimental backend is reachable from `backend="auto"` without `FLASHINFER_ALLOW_EXPERIMENTAL_AUTO_BACKENDS=1`. (Calling an `@flashinfer_experimental_api` or naming a backend explicitly is itself the opt-in and needs no environment variable.)
  - [ ] **Test scope declared below.** The experimental CI lane runs exactly these targets, so keep them as narrow as the change allows.

```experimental-tests
# One target per line: a directory or a file. (A pytest ::selector is not
# supported -- the sharding runner cannot consume one.) Must be under
# tests/experimental/ and must exist. Delete these comment lines and add yours, e.g.
#
#   tests/experimental/test_my_backend.py
#   tests/experimental/my_backend/
#
# Declaring the whole tree (tests/experimental/) is allowed but means every
# experimental PR pays for every other feature's tests, in every matrix cell.
```

## Reviewer Notes

The paged run moved to `csrc/batch_prefill_paged.cuh` in #4736. The fix is applied there too.

Generated with Claude Code and Human In The Loop 🙈

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed CUDA-graph execution for padded batch prefill and decode workloads when KV splitting is disabled.
  * Padding is now correctly identified and prevented from accessing stale workspace data.
  * Improved output consistency between CUDA-graph and non-graph execution paths.

* **Tests**
  * Added coverage for padded CUDA-graph batches using paged and ragged KV caches.
  * Strengthened validation of padding masks and workspace sizing in unsplit execution scenarios.

### CODE REVIEW

**Comment by gf239** [comment]
Rebased on main. @yzh119 @saltyminty could one of you trigger CI (`@flashinfer-bot run`) and review? The change makes the planner create the padding mask for every CUDA-graph plan, not only split-KV ones.

File: tests/attention/test_batch_decode_kernels.py [hunk 1]
```diff
@@ -1520,3 +1520,82 @@ def test_paged_decode_extreme_negative_logits(dtype):
...
+    # Fields are positional, in PrefillPlanInfo::ToVector order (scheduler.cuh).
+    info = wrapper._plan_info
+    padded_batch_size = info[0]
+    block_valid_mask_offset = info[12]
+    enable_cuda_graph, split_kv = info[13], info[14]
+    assert enable_cuda_graph and not split_kv
+    mask = wrapper._int_workspace_buffer[
+        block_valid_mask_offset : block_valid_mask_offset + padded_batch_size
+    ].bool()
+    if mask.all():
```
[the remark this answers is not in the document]
Comment (gf239) [reply tests/attention/test_batch_decode_kernels.py]: Valid, thanks. A plan that allocates the mask but never fills it did skip. The suggested `padded_batch_size == batch_size` doesn't hold, though, since the real CTA count can exceed the batch size. Fixed in ade3982, in this test and in the prefill one: the real count now comes from the plan without CUDA graphs, the test skips only when there is no padding, and the mask is compared element by element. Checked on an RTX 4050 with the mask left unfilled: the old tests skipped, the new ones fail.

**Comment by gf239** [comment]
Rebased on current main; conflicts resolved (the paged run moved to `batch_prefill_paged.cuh` in #4736). Re-verified on an RTX 4090 and an RTX 3080: the three new tests fail on main and pass with this PR, and on the 4090 the existing prefill, decode, POD and workspace-size tests give the same per-test results either way. @yzh119 @saltyminty could one of you trigger CI (`@flashinfer-bot run`) and review?

**Comment by gf239** [comment]
#5177 is merged. This is the last open fix in this planner.

Re-checked on top of current main (f17b7726) on an RTX 4090:
- the three new tests fail on main and pass with this PR;
- the prefill, decode, POD and workspace-size test files give the same per-test results with and without it.

@saltyminty could you please trigger CI (`@flashinfer-bot run`) and review?

**Comment by gf239** [comment]
@coderabbitai resume

**Comment by gf239** [comment]
@coderabbitai review

**Comment by saltyminty** [comment]
/bot run tests/attention/test_batch_prefill_kernels.py tests/attention/test_batch_decode_kernels.py tests/attention/test_workspace_size.py tests/utils/test_pod_kernels.py tests/trace/test_batch_pod_run_reference_correctness.py

**Review by saltyminty — APPROVED** [review]
Approved pending CI
